packages/server/src/adapters/fastify/fastifyRequestHandler.ts

```diff
@@ -43,7 +49,48 @@ export async function fastifyRequestHandler<
     body: opts.req.body ?? 'null',
   };
 
-  const result = await resolveHTTPResponse({
+  let resolve: (value: FastifyReply) => void;
+  const promise = new Promise<FastifyReply>((r) => (resolve = r));
+
+  const onHead = (head: HTTPResponse) => {
+    if (!opts.res.statusCode || opts.res.statusCode === 200) {
+      opts.res.statusCode = head.status;
+    }
+    for (const [key, value] of Object.entries(head.headers ?? {})) {
+      /* istanbul ignore if -- @preserve */
+      if (typeof value === 'undefined') {
+        continue;
+      }
+      void opts.res.header(key, value);
+    }
+  };
+
+  let isStream = false;
+  let stream: Readable;
+  const formatter = getBatchStreamFormatter();
+  const onChunk = ([index, string]: ResponseChunk) => {
+    if (index === -1) {
+      // full response, no streaming
+      resolve(opts.res.send(string));
+      return;
+    }
+    if (!isStream) {
+      void opts.res.header('Transfer-Encoding', 'chunked');
+      void opts.res.header(
+        'Vary',
+        opts.res.hasHeader('Vary')
+          ? 'trpc-batch-mode, ' + opts.res.getHeader('Vary')
+          : 'trpc-batch-mode',
+      );
+      stream = new Readable();
+      stream._read = () => {}; // eslint-disable-line @typescript-eslint/no-empty-function -- https://github.com/fastify/fastify/issues/805#issuecomment-369172154
+      resolve(opts.res.send(stream));
+      isStream = true;
+    }
+    stream.push(formatter(index, string));
+  };
+
+  void resolveHTTPResponse({
```

Comment:
I'm sure we shouldn't await, because of streaming, we return a Response, and leave the promise run beyond. But it still return a promise because other servers need to await the promise (Node "express-like").